### PR TITLE
STM32 GPIO interrupt controller rework

### PR DESCRIPTION
**Pull request by mathieuchopstm**
This PR is a rework of the STM32 GPIO interrupt controller API and implementation.

The main reason for this PR is the upcoming introduction of STM32WB0 series, which lack the EXTI peripheral and require these modifications for smooth introduction. However, this PR also addresses several weaknesses in the API, reduces the coupling between the GPIO and EXTI drivers, and cleans up the drivers.

### CODE REVIEW

**Comment by mathieuchopstm** [comment]
Added missing `#include`s causing build failures on STM32F1 and STM32L0 series (should fix CI failures)

File: drivers/interrupt_controller/intc_exti_stm32.c [hunk 1]
```diff
@@ -34,130 +39,85 @@ struct stm32_exti_range {
...
-	/* Enable exti irq interrupt */
-	irq_enable(irqnum);
-}
-
-void stm32_exti_disable(int line)
+/**
+ * @returns the LL_<PPP>_EXTI_LINE_xxx define that corresponds to specified @p linenum
+ * This value can be used with the LL EXTI source configuration functions.
+ */
+static inline uint32_t stm32_exti_linenum_to_src_cfg_line(gpio_pin_t linenum)
```
Comment (erwango) [inline drivers/interrupt_controller/intc_exti_stm32.c:58]: Non blocking nitpick: `line_num` would be more readable than `linenum`

**Review by erwango — DISMISSED** [review]
A welcome and nice change, thanks.

A follow up change can introduce `clocks` prop in exti node to make it easier to introduce new series.

**Comment by mathieuchopstm** [comment]
`#include` the correct header to fix build failures on STM32L0 (`<zephyr/dt-bindings/gpio/stm32-gpio.h>` was not the correct one...)

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Pull request merged**